File: LeetCode/320-generalized-abbreviation/ex_320.cpp
class Solution {
public:
    vector<string> generateAbbreviations(string word) {
        vector<string> res;
        helper(res, word, "", 0, true);
        return res;
    }
private:
    void helper(vector<string> &res, string word, string abbr, int i, bool use_number) {
        if (i == word.length()) {
            res.push_back(abbr);
            return;
        }
        helper(res, word, abbr + word[i], i + 1, true);
        if (use_number) {
            for (int j = 1; i + j <= word.length(); ++j) {
                helper(res, word, abbr + to_string(j), i + j, false);
            }
        }
    }
};

File: LeetCode/320-generalized-abbreviation/ex_320.java
public class Solution {
    public List<String> generateAbbreviations(String word) {
        List<String> res = new ArrayList<>();
        helper(res, word, "", 0, true);
        return res;
    }
    
    private void helper(List<String> res, String word, String abbr, int i, boolean useNumber) {
        if (i == word.length()) {
            res.add(abbr);
            return;
        }
        helper(res, word, abbr + word.charAt(i), i + 1, true);
        if (useNumber) {
            for (int j = 1; i + j <= word.length(); j++) {
                helper(res, word, abbr + Integer.toString(j), i + j, false);
            }
        }
    }
}